### PR TITLE
Added SearchLambdaCollection utility

### PR DESCRIPTION
**Pull request by adamyy**
* Per our discussion in the slack channel, the search lambda now only returns the docids and the QA lambda needs to query from DynamoDB, in order to reduce the total bandwidth usage
* Added SearchLambdaCollection (analogous to SearchCollection in Anserini, but for Lambda)

A test run with [msmarco-passage](https://github.com/castorini/anserini/blob/master/docs/regressions-msmarco-passage.md) gave me a total run time of roughly 9 minutes for 6980 queries on a EC2 machine in the same region. For comparison, the same search takes about 5 minutes on my personal machine (8-core Intel Core i9). The delay/duration metric collected by Lambda is shown below. TL;DR, most queries (p95) finishes under a second, and the average is generally around half a second.

image

@lintool @shaneding

### CODE REVIEW

**Comment by adamyy** [comment]
@lintool Lambda does not provide me with many knobs that I can tune regarding concurrency. The two configurable numbers are **reserved concurrency** and **provisioned concurrency**. 

Reserved concurrency basically tells AWS to reserve a fixed number of capacity units so that lambda can scale up to that without being interrupted by other lambda functions. In my understanding, each AWS account has a fixed number of account concurrency units of 1000 (per region), and if you don't reserve account concurrency for a function, your other functions can eat up those capacity.

Provisioned concurrency tells AWS to keep a number of lambda containers to stay warm and ready, basically AWS's recommendation for resolving cold start. But they charges you extra for it by hours and units of usage (and they are quite costly).
image

**Comment by adamyy** [comment]
> For comparison, the same search takes about 5 minutes on my personal machine (8-core Intel Core i9). 

Actually not the same search. `SearchCollection` also does some slightly more work regarding with its default bm25 settings (e.g., rerank with `ScoreTiesAdjusterReranker`).

File: search-lambda-function/src/main/java/io/anlessini/SearchRequest.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  @Override
+  public String toString() {
+    return "SearchRequest{" +
+        "query='" + query + '\'' +
+        ", maxDocs=" + maxDocs +
+        ", bm25k1=" + bm25k1 +
+        ", bm25b=" + bm25b +
+        '}';
```
Comment (lintool) [inline search-lambda-function/src/main/java/io/anlessini/SearchRequest.java:67]: double quotes to be consistent?

File: search-lambda-function/src/main/java/io/anlessini/SearchResponse.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+          ", doc=" + doc +
+          '}';
+    }
+  }
+
+  @Override
+  public String toString() {
+    return "SearchResponse{" +
+        "hits=" + hits +
+        '}';
```
Comment (lintool) [inline search-lambda-function/src/main/java/io/anlessini/SearchResponse.java:45]: same

**Review by lintool — APPROVED** [review]
lgtm

**Comment by adamyy** [comment]
Hi @lintool, I tested the throughput multiple times to eliminate the effect of cold start.

```bash
utils/target/appassembler/bin/SearchLambdaCollection -lambda $LAMBDA_ARN \
-topics topics.msmarco-passage.dev-subset.txt -topic.reader TsvInt \
-output run.msmarco-passage.bm25-default.topics.msmarco-passage.dev-subset.txt
```

||8 Client Threads|16 Client Threads|
|:---|:---|:---|
|Total run time|00:08:23|00:04:12|

The 16-threads config does see a 2x throughput given lambda is not throttled in this case.

Hi @shaneding, did you have the time to briefly go over this? Basically the changes related to you is that your QA/Rerank lambda needs to invoke the search lambda for docids with the SDK rather than from the API endpoint (see [boto3](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/lambda.html#Lambda.Client.invoke) doc), then it would use docids to query Dynamo for doc contents.

image

I will reflect these changes in #21 after this PR is merged.

**Comment by adamyy** [comment]
Merging this would break the current behavior of the project, as the Lucene lambda now only returns docids but not document contents, which is not what the frontend expect.

As a next step, to make the frontend happy, I will create a simple Python lambda in the place of the QA/Reranking lambda. It would be connected to the API gateway, invoke the Lucene lambda, then enrich the search results with document contents from DynamoDB before returning it. This should restore the previous behavior. Later, this Python lambda can be modified to run reranking/QA model if desired.

**Pull request merged**